pkg/pillar/cmd/nodeagent/nodeagent.go

```diff
@@ -0,0 +1,484 @@
+
+	// start the watchdog process timer tick
+	stillRunning := time.NewTicker(25 * time.Second)
+
+	// Make sure we have a GlobalConfig file with defaults
+	types.EnsureGCFile()
+
+	// publisher of NodeAgent Status
+	pubNodeAgentStatus, err := pubsub.Publish(agentName, types.NodeAgentStatus{})
+	if err != nil {
+		log.Fatal(err)
+	}
+	pubNodeAgentStatus.ClearRestarted()
+	nodeagentCtx.pubNodeAgentStatus = pubNodeAgentStatus
+
+	// publisher of Zboot Config
+	pubZbootConfig, err := pubsub.Publish(agentName, types.ZbootConfig{})
+	if err != nil {
+		log.Fatal(err)
+	}
+	pubZbootConfig.ClearRestarted()
+	nodeagentCtx.pubZbootConfig = pubZbootConfig
+
+	// Look for global config such as log levels
+	subGlobalConfig, err := pubsub.Subscribe("", types.GlobalConfig{},
+		false, &nodeagentCtx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	subGlobalConfig.ModifyHandler = handleGlobalConfigModify
+	subGlobalConfig.DeleteHandler = handleGlobalConfigDelete
+	nodeagentCtx.subGlobalConfig = subGlobalConfig
+	subGlobalConfig.Activate()
+
+	// monitor device status through led blinks
+	subLedBlinkConfig, err := pubsub.Subscribe("", types.LedBlinkCounter{},
+		false, &nodeagentCtx)
+	if err != nil {
+		log.Fatal(err)
+	}
+	subLedBlinkConfig.ModifyHandler = handleLedBlinkConfigModify
+	subLedBlinkConfig.DeleteHandler = handleLedBlinkConfigDelete
+	nodeagentCtx.subLedBlinkConfig = subLedBlinkConfig
+	subLedBlinkConfig.Activate()
+
+	// baseline timers
+	nodeagentCtx.lastConfigReceivedTime = time.Now()
+
+	// publish zboot config as of now
+	publishZbootConfigAll(&nodeagentCtx)
+
+	// access the zboot APIs directly, baseosmgr is still not ready
+	nodeagentCtx.updateInprogress = zboot.IsCurrentPartitionStateInProgress()
+	log.Infof("Current partition: %s, inProgress: %v\n", nodeagentCtx.curPart,
+		nodeagentCtx.updateInprogress)
+	log.Infof("configReceivedTime:%v\n", nodeagentCtx.lastConfigReceivedTime)
+
+	// Read the GlobalConfig first
+	// Wait for initial GlobalConfig
+	log.Infof("Waiting for GCInitialized\n")
+	for !nodeagentCtx.GCInitialized {
+		select {
+		case change := <-subGlobalConfig.C:
+			subGlobalConfig.ProcessChange(change)
+
+		case change := <-subLedBlinkConfig.C:
+			subLedBlinkConfig.ProcessChange(change)
+
+		case <-stillRunning.C:
+			agentlog.StillRunning(agentName)
+			handleDeviceTimers(&nodeagentCtx)
+		}
+	}
+
+	// if current partition state is not in-progress,
+	// nothing much to do. Zedcloud connectivity is tracked,
+	// to trigger the device to reboot, on reset timeout
+	//
+	// if current partition state is in-progress,
+	// trigger the device to reboot on
+	// fallback timeout period, if the zedbox modules
+	// come up in this time period,
+	//
+	// On zedbox modules activation, nodeagent will
+	// track the zedcloud connectivity events
+	//
+	// These timer functions will be tracked using
+	// led blinker and cloud connectionnectivity status.
+
+	// baseline timers, again
+	nodeagentCtx.lastConfigReceivedTime = time.Now()
```

Comment:
I suspect you should leave this at zero and then check if it has been set by zedagent.
Note that deviceagent starts before ntp, hence your time could be the year 1970 or 2016.
Recording a timestamp would fail on device which doesn't have a hardware clock.